yjit_codegen.c

```diff
@@ -1900,18 +1904,31 @@ gen_branchif(jitstate_t* jit, ctx_t* ctx)
         yjit_check_ints(cb, side_exit);
     }
 
-    // Test if any bit (outside of the Qnil bit) is on
-    // RUBY_Qfalse  /* ...0000 0000 */
-    // RUBY_Qnil    /* ...0000 1000 */
-    x86opnd_t val_opnd = ctx_stack_pop(ctx, 1);
-    test(cb, val_opnd, imm_opnd(~Qnil));
-
     // Get the branch target instruction offsets
     uint32_t next_idx = jit_next_insn_idx(jit);
     uint32_t jump_idx = next_idx + jump_offset;
     blockid_t next_block = { jit->iseq, next_idx };
     blockid_t jump_block = { jit->iseq, jump_idx };
 
+    // Check if we know enough from type info
+    val_type_t val_type = ctx_get_opnd_type(ctx, OPND_STACK(0));
+    if (val_type.type != ETYPE_UNKNOWN) {
+        ctx_stack_pop(ctx, 1);
```

Comment:
I think you would have to pop the value regardless of whether its type is known or not.

Reply:
There was a pop lower if we didn't hit this case, but they probably should have been combined for clarity.